Start Video Stream With OnTouchEvent Permission (#947)
smartdevicelink/sdl_evolution: proposals/nnnn-stream-ontouchevent.md

# Start Video Stream With OnTouchEvent Permission

* Proposal: [SDL-NNNN](nnnn-stream-ontouchevent.md)
* Author: [Kujtim Shala](https://github.com/kshala-ford)
* Status: **Awaiting review**
* Impacted Platforms: [ iOS ]

## Introduction

> This proposal is changing the behavior how the streaming video manager should include `OnTouchEvent` permission into the starting conditions.

Reviewer: ```suggestion
This proposal is changing the behavior for how the streaming video manager should include `OnTouchEvent` permission in the starting conditions.
```

## Motivation

> SDL navigation applications stream the UI as a video to the vehicle head unit. In this case it's quite obvious that touch events are key for such applications to operate. However, for nav apps that connect to a head unit for the first time, the app ID is unknown and therefore it's not clear if the app is allowed to video stream or to receive touch events. Starting the video stream only upon `HMI_FULL` can lead to Core to reply with `NAK` on the start service request.

Reviewer: ```suggestion
SDL navigation applications stream the UI as a video to the vehicle head unit. In this case it's quite obvious that touch events are key for such applications to operate. However, for nav apps that connect to a head unit for the first time, the app ID is unknown and therefore it's not clear if the app is allowed to video stream or to receive touch events. Starting the video stream only upon `HMI_FULL` can lead to Core to replying with `NAK` on the start service request.
```

## Proposed solution

The library should wait to send the start service request until all conditions are met. The solution is to check the existence of `OnTouchEvent` in the latest `OnPermissionsChange` notification.

Below code shows an example implementation in the class `SDLStreamingVideoLifecycleManager`. 

> First of all the manager should be able to store the last known permission status of OnTouchEvent in case that the permissions were provided before other conditions were met (e.g. permission update during HMI_NONE).

Reviewer: ```suggestion
First of all the manager should be able to store the last known permission status of OnTouchEvent in case the permissions were provided before other conditions were met (e.g. permission update during HMI_NONE).
```

        [self sdl_startVideoSession];
    } else {
        [self sdl_stopVideoSession];
    }
}
```

With the newly added flag, every other place that starts or stops the video session should be altered as well, including the following:

```objc
- (void)startWithProtocol:(SDLProtocol *)protocol {
     // attempt to start streaming since we may already have necessary conditions met
-    [self sdl_startVideoSession];
+    if (self.isHmiStateVideoStreamCapable && self.videoStreamPermissionGranted) {
+        [self sdl_startVideoSession];
+    } else {
+        [self sdl_stopVideoSession];
+    }
}

- (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
    ...
-    if (self.isHmiStateVideoStreamCapable) {
+    if (self.isHmiStateVideoStreamCapable && self.videoStreamPermissionGranted) {
         [self sdl_startVideoSession];
     } else {
         [self sdl_stopVideoSession];
     }
}
```

Note: The example code should not be considered as the only option to implement this proposal. Other solutions that serve the same feature should also be accepted during implementation.  

## Potential downsides

> Potentially this code can have an impact on existing head units that don't know if navigation apps start the video stream. Those applications would require OnTouchEvent anyway for the app to operate, so they shouldn't see any impact. Navigation applications could be delayed by that. However as head units may return `NAK` in case of an early start service request this proposal should rather be seen as a stability improvement.

Author: ```suggestion
Potentially this code can have an impact on existing head units that would allow navigation apps to start the video stream without touch permissions. Those applications require OnTouchEvent anyway to operate, so they shouldn't see any impact. Navigation applications could be delayed, however, as head units may return `NAK` in case of an early start service request this proposal should rather be seen as a stability improvement.
```

Author: @theresalech Please see my revised suggestion.

Reviewer: Sorry, I missed this earlier!  This looks good to me, with the following change:
Current: `Navigation applications could be delayed...`
New: `Navigation applications' startup time could be delayed...`

> Potentially this code can have impact to existing head units that don't mind with navigation apps to start the video stream. Those applications would require OnTouchEvent anyways for the app to operate hence they shouldn't see any impact. Navigation applications could be delayed by that. However as head units may return `NAK` in case of an early start service request this proposal should be rather seen as a stability improvement.

Reviewer: I am a little unclear on your intended meaning here - specifically regarding the first sentence - so please let me know if my suggested revisions incorrectly change what you're trying to say.  Thanks!

```suggestion
Potentially this code can have an impact on existing head units that don't know if navigation apps start the video stream. Those applications would require OnTouchEvent anyway for the app to operate, so they shouldn't see any impact. Navigation applications could be delayed by that. However as head units may return `NAK` in case of an early start service request this proposal should rather be seen as a stability improvement.
```

Author: I'm sorry for the ambiguity. Touch permissions are not technically required to start a video stream but we imply both to be necessary for navigation apps. 

If a head unit allows apps to start the video stream but touch isn't granted those apps would be basically broken. Touch events would not be forwarded to the app, therefore it can't react per user input.

Reviewer: @kshala-ford thanks for the clarification!  I've revised my suggested change in the comment above.  Please commit this change if it makes sense, or feel free to revise as needed.

## Impact on existing code

Only the video streaming manager is in focus. Ideas to allow navigation applications without streaming shouldn't be affected as they wouldn't start the streaming managers.

## Alternatives considered

> The video streamable status notification should provide a very similar functionality, however it's not available on all nav-app capable head units therefore an alternative was considered listening for touch events. Also listening for the video streamable status could be a valid addition to the proposed solution.

Reviewer: ```suggestion
The video streamable status notification should provide very similar functionality, however it's not available on all nav app-capable head units and therefore an alternative was considered listening for touch events. Also listening for the video streamable status could be a valid addition to the proposed solution.
```
